pkg/server/registry/apigroups/acorn/jobs/restart.go

```diff
@@ -0,0 +1,112 @@
+	"k8s.io/apiserver/pkg/registry/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewRestart(c client.WithWatch) rest.Storage {
+	return stores.NewBuilder(c.Scheme(), &apiv1.JobRestart{}).
+		WithCreate(&restartStrategy{client: c}).
+		WithValidateName(validator.NoValidation).
+		Build()
+}
+
+type restartStrategy struct {
+	client client.WithWatch
+}
+
+func (s *restartStrategy) New() types.Object {
+	return &apiv1.JobRestart{}
+}
+
+func (s *restartStrategy) Create(ctx context.Context, obj types.Object) (types.Object, error) {
+	ri, _ := request.RequestInfoFrom(ctx)
+
+	if ri.Namespace == "" || ri.Name == "" {
+		return obj, nil
+	}
+
+	// Find the job to restart by looking it up with the name and namespace from the request info.
+	job := &apiv1.Job{}
+	err := s.client.Get(ctx, kclient.ObjectKey{Namespace: ri.Namespace, Name: ri.Name}, job)
+	if err != nil {
+		return nil, err
+	}
+
+	key := kclient.ObjectKey{Namespace: job.Status.JobNamespace, Name: job.Status.JobName}
+
+	cronToRestart := &batchv1.CronJob{}
+	if err = s.client.Get(ctx, key, cronToRestart); err == nil {
+		return obj, s.restartCronJob(ctx, cronToRestart)
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return obj, err
+	}
+
+	jobToRestart := &batchv1.Job{}
+	if err = s.client.Get(ctx, key, jobToRestart); err != nil {
+		return obj, err
+	}
+
+	// Delete the Job and set the propagation policy to foreground so that the dependent resources (Pods)
+	// created by the Job are also deleted.
+	opts := &client.DeleteOptions{PropagationPolicy: z.Pointer(metav1.DeletePropagationForeground)}
+	return obj, s.client.Delete(ctx, jobToRestart, opts)
+}
+
+func (s *restartStrategy) restartCronJob(ctx context.Context, cron *batchv1.CronJob) error {
+	// Find all active jobs and delete them to make way for new ones.
+	for _, jobRef := range cron.Status.Active {
+		job := &batchv1.Job{}
+		err := s.client.Get(ctx, kclient.ObjectKey{Name: jobRef.Name, Namespace: jobRef.Namespace}, job)
+		if err != nil {
+			return err
+		}
+
+		// Delete the Job and set the propagation policy to foreground so that the dependent resources (Pods)
+		// created by the Job are also deleted.
+		opts := &client.DeleteOptions{PropagationPolicy: z.Pointer(metav1.DeletePropagationForeground)}
+		if err := s.client.Delete(ctx, job, opts); err != nil {
+			return err
+		}
+	}
+
+	// Create a new job from the cron's template to "restart" it. This is necessary because
+	// the cronjob controller will not create a new job until the next scheduled run. By creating
+	// it here we can ensure that the job is restarted immediately.
+	template := cron.Spec.JobTemplate
+	spec := template.Spec
+	spec.TTLSecondsAfterFinished = new(int32) // Want to delete the job immediately after it finishes.
+	newJob := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: cron.Name + "-",
+			Namespace:    cron.Namespace,
+			Labels:       template.Labels,
+			Annotations:  template.Annotations,
```

Comment:
Does this match the behavior of the k8s controller for CronJobs? Does it just copy labels and annotations from the CronJob to the Job it creates for it? Idk whether it does or not but I think we should make sure we are consistent with that behavior.